contracts/external/cw-vesting/src/contract.rs

```diff
@@ -0,0 +1,556 @@
+    }
+
+    // Check vesting payment denom matches local denom
+    let denom = match vesting_payment.denom {
+        CheckedDenom::Cw20(_) => {
+            return Err(ContractError::NotStakeable);
+        }
+        CheckedDenom::Native(denom) => {
+            // Get local denom, ensure it matches the vesting denom
+            let local_denom = deps.querier.query_bonded_denom()?;
+            if local_denom != denom {
+                return Err(ContractError::NotStakeable);
+            } else {
+                denom
+            }
+        }
+    };
+
+    // Load source validator, subtract redelegated amount
+    let amount_staked_src = STAKED_VESTING_BY_VALIDATOR
+        .load(deps.storage, &src_validator)
+        .map_err(|_| ContractError::NoDelegationsForValidator {})?;
+
+    STAKED_VESTING_BY_VALIDATOR.save(
+        deps.storage,
+        &src_validator,
+        &amount_staked_src.checked_sub(amount)?,
+    )?;
+
+    // Load destination validator, increase amount
+    match STAKED_VESTING_BY_VALIDATOR.may_load(deps.storage, &dst_validator)? {
+        // If already staked to this validator, increase staked amount
+        Some(staked_amount) => STAKED_VESTING_BY_VALIDATOR.save(
+            deps.storage,
+            &src_validator,
+            &staked_amount.checked_add(amount)?,
+        ),
+        // If not currently staked to this validator save a new record with staked amount
+        None => STAKED_VESTING_BY_VALIDATOR.save(deps.storage, &dst_validator, &amount),
+    }?;
+
+    // Create message to delegate the underlying tokens
+    let msg = StakingMsg::Redelegate {
+        src_validator,
+        dst_validator,
+        amount: Coin { denom, amount },
+    };
+
+    Ok(Response::new()
+        .add_attribute("method", "redelegate")
+        .add_attribute("amount", amount.to_string())
+        .add_message(msg))
+}
+
+pub fn execute_undelegate(
+    _env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    validator: String,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    let vesting_payment = VESTING_PAYMENT.load(deps.storage)?;
+
+    // Check sender is the recipient of the vesting payment
+    if info.sender != vesting_payment.recipient {
+        return Err(ContractError::Unauthorized);
+    }
+
+    // Update vesting payment with amount of staked rewards
+    VESTING_PAYMENT.update(deps.storage, |mut vp| -> Result<_, ContractError> {
+        // Check unstake amount is not greater than vesting payment amount
+        if amount > vp.staked_amount {
+            return Err(ContractError::NotEnoughFunds {});
+        }
+
+        // Update amounts
+        vp.staked_amount = vp.staked_amount.checked_sub(amount)?;
+        Ok(vp)
+    })?;
+
+    let denom = deps.querier.query_bonded_denom()?;
+
+    // Create message to delegate the underlying tokens
+    let msg = StakingMsg::Undelegate {
+        validator: validator.clone(),
+        amount: Coin { denom, amount },
+    };
```

Comment:
what happens once this undelegation completes? does the rewards receiver get the tokens?

seeing as we don't re-increment `vp.amount` here once undelegation completes, i assume yes? in which case i feel i could trivially bypass the vesting by staking all the tokens, and then immediately undelegating them.

Reply:
> if this contract receives the undelegated funds, shouldn't we be checking the status of unbonding funds in execute_distribute to make sure we don't just error with 'no tokens' if someone distributes and their tokens are in the unbonding or staked state?

We could do this, but it's actually surprisingly difficult. `cosmwasm_std` does not have support for querying unbonding delegations yet. The delegations query doesn't return any info about it: https://docs.rs/cosmwasm-std/latest/cosmwasm_std/struct.FullDelegation.html

> if the rewards receiver address receives these unbonded tokens, then i feel it is trivial to bypass vesting entirely by bonding and unbonding the tokens.

It does not, the unbonded tokens need to go to the contract since they may not have fully vested.

While we could return a more helpful error message, I feel the current contract is perhaps "good enough" here as the UI could more than compensate for this by showing tokens are unbonding and that there is nothing to claim.

Reply:
> I feel the current contract is perhaps "good enough" here as the UI could more than compensate for this by showing tokens are unbonding and that there is nothing to claim.

thanks for the detailed explanation. i agree with this take. i also recall this being a pain to query from a conversation with @NoahSaso about the DAO treasury cards. validating this in the FE sounds good to me.